### PR TITLE
reminder + tz + user cmd

### LINKED ISSUE #39: Timezone (reminders)
No support for timezones, perhaps per guild.
https://discord.com/channels/780572565240414208/780572565240414211/785574409989128222

### LINKED ISSUE #55: user dm not handled
can't set reminders via dm

### PR DESCRIPTION
**Pull request by PandaGamersHDNL**

### CODE REVIEW

File: src/commands.ts [hunk 1]
```diff
@@ -10,52 +10,64 @@ import { WikiService } from './services/wiki-service';
...
-      };
-      return embed;
+      } as MessageEmbedOptions;
+
     }
   },
   { // setup
     name: 'setup',
-    description: 'Quick setup and introduction for the bot',
+    description: 'Quick setup and introduction for the bot. server only',
```
Comment (DrunkenToast) [inline src/commands.ts]: Typos

File: src/commands.ts [hunk 2]
```diff
@@ -191,10 +212,13 @@ export const commands: Command[] = [
   },
   { // prefix
     name: 'prefix',
-    description: 'Set prefix for guild',
+    description: 'Set prefix for guild. server only',
```
Comment (DrunkenToast) [inline src/commands.ts]: Capital S and .

File: src/commands.ts [hunk 3]
```diff
@@ -191,10 +212,13 @@ export const commands: Command[] = [
...
     aliases: ['pf'],
     async response(msg: Message, guildConfig: GuildConfig): Promise<Response | void> {
-      const tokenizer = new Tokenizer(msg.content, guildConfig);
+      if (!guildConfig) {
+        return 'this is a guild only command';
```
Comment (DrunkenToast) [inline src/commands.ts]: T and .

File: src/commands.ts [hunk 4]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
       } else {
         return 'this is not a valid channel';
       }
     }
-  },
+  },*/
   { // notes
     name: 'notes',
-    description: 'set or get notes for channels',
+    description: 'set or get notes for channels. server only',
```
Comment (DrunkenToast) [inline src/commands.ts]: S, S and .

File: src/commands.ts [hunk 5]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
     aliases: ['note'],
     async response(msg: Message, guildConfig: GuildConfig): Promise<Response | void> {
-      const tokenizer = new Tokenizer(msg.content, guildConfig);
+      if (!guildConfig) {
+        return 'this is a server command only';
```
Comment (DrunkenToast) [inline src/commands.ts]: S and .

File: src/commands.ts [hunk 6]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
+            ReminderService.create({
+              content: tokenizer.body(3),
+              date: time.toISO(),
+              target: {
+                user: msg.author!.id
+              },
+            });
+          }
+          const userTime = time.setZone(await ReminderService.getTimeZone(msg.author.id), { keepLocalTime: true });
+          return 'your reminder has been set as: ' + userTime.toString();
```
Comment (DrunkenToast) [inline src/commands.ts]: Y

File: src/commands.ts [hunk 7]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
+        return {
+          'title': 'clock',
+          'description': '`clock get`: gets you your time\n `clock set`: set your clock',
+          'color': '4FAFEF',
+        } as MessageEmbedOptions;
+      if (tokenizer.tokens[1].content == 'get') {
+        const tz = (await ReminderService.getTimeZone(msg.author.id));
+        console.log(tz);
+        const time = DateTime.fromJSDate(new Date(), { zone: tz });
+        return `bot thinks it's ${time.toString()} for you with time zone ${time.zoneName}`;
```
Comment (DrunkenToast) [inline src/commands.ts:358]: B and .

File: src/commands.ts [hunk 8]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
+          const time = DateTime.fromMillis(Date.now(), { zone: tz });
+
+          if (time.isValid) {
+            await ReminderService.setTimeZone(msg.author.id, tz);
+            return `bot thinks it's ${time.toString()} for you with time zone ${time.zoneName}`;
+          }
+        }
+        let i = 0;
+        return {
+          'title': 'time zones!',
```
Comment (DrunkenToast) [inline src/commands.ts]: Timezones

File: src/commands.ts [hunk 9]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
+        if (tokenizer.tokens[2]) {
+          let tz = timeZones[parseInt(tokenizer.tokens[2].content)];
+          if (!tz) {
+            tz = tokenizer.tokens[2].content;
+          }
+          const time = DateTime.fromMillis(Date.now(), { zone: tz });
+
+          if (time.isValid) {
+            await ReminderService.setTimeZone(msg.author.id, tz);
+            return `bot thinks it's ${time.toString()} for you with time zone ${time.zoneName}`;
```
Comment (DrunkenToast) [inline src/commands.ts:370]: B and .

File: src/commands.ts [hunk 10]
```diff
@@ -209,97 +233,161 @@ export const commands: Command[] = [
...
+        return {
+          'title': 'time zones!',
+          'url': 'https://en.wikipedia.org/wiki/List_of_tz_database_time_zones',
+          'description': timeZones.map(tz => `\`${i++}:\`${tz}`).join('\n') + '\n',
+          'color': '43B581',
+        } as MessageEmbedOptions;
+      }
+      return {
+        'title': 'clock',
+        'description': 'clock get: gets you your time\n clock set: set your clock',
```
Comment (DrunkenToast) [inline src/commands.ts]: C

File: src/commands.ts [hunk 11]
```diff
@@ -324,25 +412,11 @@ export const commands: Command[] = [
   },
   { // courses menu command
     name: 'courses',
-    description: 'Lists your courses, modules and items with controls',
+    description: 'Lists your courses, modules and items with controls. guild command',
```
Comment (DrunkenToast) [inline src/commands.ts:419]: G and .

File: src/discord.ts [hunk 12]
```diff
@@ -44,11 +44,33 @@ export async function buildClient(shard: number, shardCount: number): Promise<Cl
...
+      if (!tokenizer.command()) {
+        return; // not a valid command
+      }
+
+      for (const command of commands) {
+        if (tokenizer.command() !== command.name && !command.aliases.includes(tokenizer.command()!)) {
+          continue;
+        }
+  
+        Logger.debug(`received dm command '${tokenizer.command()}' from user ${msg.author.id}`);
```
Comment (DrunkenToast) [inline src/discord.ts]: R and .

File: src/services/reminder-service.ts [hunk 13]
```diff
@@ -24,9 +25,8 @@ export class ReminderService {
 
   static sendGuildReminder(reminder: GuildReminder, client: Client): void {
     try {
-
       (client.channels.resolve(reminder.target.channel) as TextChannel)
-        .send(reminder.content);
+        .send(reminder.content || 'reminder');
```
Comment (DrunkenToast) [inline src/services/reminder-service.ts]: Just 'reminder' seems a bit bland maybe?

File: src/services/reminder-service.ts [hunk 14]
```diff
@@ -62,4 +62,20 @@ export class ReminderService {
       url: `/reminders/${userID}/${lastAssignment}`,
     });
   }
+
+  static async getTimeZone(discordID: string): Promise<string> {
```
Comment (DrunkenToast) [inline src/services/reminder-service.ts:67]: getTimezone

**Review by DrunkenToast — CHANGES_REQUESTED** [review]
Nitpicks

File: src/services/reminder-service.ts (hunk 13 above)
Comment (PandaGamersHDNL) [reply src/services/reminder-service.ts]: suggest something else.
maybe time of reminder

Comment (DrunkenToast) [reply src/services/reminder-service.ts]: Mention the user with 'here's your reminder for {time}.' or something like that.

**Review by xunafay — APPROVED** [review]

**Review by DrunkenToast — APPROVED** [review]

**Pull request merged**